content/unroll-default-arguments.md

```diff
@@ -0,0 +1,817 @@
+      allowRepeats: Boolean = false,
+      totalWidth: Int = 100,
+      printHelpOnExit: Boolean = true,
+      docsOnNewLine: Boolean = false,
+      autoPrintHelpAndExit: Option[(Int, PrintStream)] = Some((0, System.out)),
+      customName: String = null,
+      customDoc: String = null,
+      @unroll sorted: Boolean = true,
+      nameMapper: String => Option[String] = Util.kebabCaseNameMapper
+  ): Either[String, T] = ???
+```
+
+This allows the developer to write the minimal amount of code they _want_ to write,
+and add a single annotation to allow binary compatibility to old versions. In this
+case, we annotated `sorted` with `@unroll`, which generates forwarders that make
+`def constructEither` binary compatible with older versions that have fewer parameters,
+up to a version before `sorted` was added. Any existing method `def`, `class`, or
+`case class` can be evolved in this way, by addition of `@unroll` the first time
+a default argument is added to their signature after its initial definition.
+
+### Unrolling `def`s
+
+Consider a library that is written as follows:
+
+```scala
+object Unrolled{
+   def foo(s: String, n: Int = 1) = s + n + b + l
+}
+```
+
+If over time a new default parameter is added:
+
+```scala
+object Unrolled{
+   def foo(s: String, n: Int = 1, b: Boolean = true) = s + n + b + l
+}
+```
+
+And another
+
+```scala
+object Unrolled{
+   def foo(s: String, n: Int = 1, b: Boolean = true, l: Long = 0) = s + n + b + l
+}
+```
+
+This is a source-compatible change, but not binary-compatible: JVM bytecode compiled against an 
+earlier version of the library would be expecting to call `def foo(String, Int)`, but will fail
+because the signature is now `def foo(String, Int, Boolean)` or `def foo(String, Int, Boolean, Long)`.
+On the JVM this will result in a `MethodNotFoundError` at runtime, a common experience for anyone
+who upgrading the versions of their dependencies. Similar concerns are present with Scala.js and
+Scala-Native, albeit the failure happens at link-time rather than run-time
+
+`@unroll` is an annotation that can be applied as follows, to the first "additional" default
+parameter that was added (in this case, `b: Boolean = true`)
+
+
+```scala
+import scala.annotation.unroll
+
+object Unrolled{
+   def foo(s: String, n: Int = 1, @unroll b: Boolean = true, l: Long = 0) = s + n + b + l
+}
+```
+
+The `@unroll` annotation takes `def foo` and generates synthetic forwarders for the purpose
+of maintaining binary compatibility for old callers who may be expecting the previous signature.
+These forwarders do nothing but forward the call to the current implementation, using the
+given default parameter values:
+
+```scala
+import scala.annotation.unroll
+
+object Unrolled{
+   def foo(s: String, n: Int = 1, @unroll b: Boolean = true, l: Long = 0) = s + n + b + l
+
+   def foo(s: String, n: Int, b: Boolean) = foo(s, n, b, 0)
+   def foo(s: String, n: Int) = foo(s, n, true, 0)
```

Comment:
This is actually a bit simplified. We do not inline the constants `true` and `0`, but instead get them from the `this.foo$default$1` and `this.foo$default$2` methods. This means that if someone overrides the default with a different, the overriden value is used. But for the vast majority of scenarios that does not make a difference to the user

Reply:
Understood. But why do we need two generated `foo` functions? I would have thought if we add a new `b` parameter, we put the unroll on the `b` and get:

```scala
def foo(s: String, n: Int) = foo(s, n, true, 0)
```
Isn't that enough? Why do we need the three parameter `foo`?

Reply:
The assumption was tha new default parameters were added one at a time, so two/three/four parameter versions of `foo` exist and were published, and so we need the corresponding forwarders for code compiled against each version

If that assumption is not true, then `@unrollOnly` gives more fine grained support on what forwarders need to be generated

Reply:
I still don't get it. Can you walk me through step by step. The way I see this is:

Version 1 has
```scala
def foo(s: String, n: Int = 0)
```
I threw in a default, just to cover that case.

Version 2 has:
```scala
def foo(s: String, n: Int = 0, @unroll b: Boolean = true, l: Long = 0)
def foo(s: String, n: Int) = foo(s, n, true, 0) // auto-generated
```
The default parameter on the generated version is gone. But that's OK because we have the 
same default accessors, so old code calling `foo("hello")` expanded to `foo("hello", foo$default$2)` and that also works fine against the new generated version. 

Version 3 has:
```scala
def foo(s: String, n: Int = 0, @unroll b: Boolean = true, l: Long = 0, @unroll f: Float = 0.0f), c: Char = 'a')
def foo(s: String, n: Int., b: Boolean, l: Long) = foo(s, n, b, l, 0.0f, 'a')  // auto-generated
def foo(s: String, n: Int) = foo(s, n, true, 0) // auto-generated
```

I must be missing something, since there's always exactly one function generated, but I don't see what it is.

Reply:
What you described is exactly the (as-currently-written) semantics for `@unrollOnly`. You are not missing anything,  it's correct. With 3 versions, adding two default arguments in versions 2 and 3, only 2 generated forwarders are necessary to maintain binary compatibility.

I think where your understanding differs from the current proposal, is that the current proposal calls what you described `@unrollOnly` rather than `@unroll`. In the current proposal, where `@unroll` is shorthand for _"add `@unrollOnly` to the parameter annotated and every parameter to the right"_.

The naming should definitely be changed if it's unintuitive. If you and @bjornregnell both find the current naming confusing, that's probably the case haha

Reply:
Ah, but why would you use `unroll` then over `unrollOnly`?  If you add multiple parameters in one version, there's no need to add a new forwarder for each prefix. I think I would prefer just one annotation. Call it `unroll` with the semantics of `unrollOnly`. Then if you want multiple forwarders, just put multiple @unrolls on your parameters. That's not a big burden, and would be less confusing.

Reply:
I have updated the document to use `@unroll`/`@unrollAll`

I have not updated the prototype implementation, but it should be easy to do so when the time comes

Reply:
Or, if we go with an annotation on the method, it could be 
```scala
@unrollFrom("n") 
def foo(s: String, n: Int = 0, b: Boolean = true)
```
vs
```scala
@unrollAt("n") 
def foo(s: String, n: Int = 0, b: Boolean = true)
```

(Now it's me who caught the naming bug)

Reply:
Yes, that works, but I would argue that having a secondary annotation on the parameter itself is much better than duplicating the parameter name above the method def. 

A secondary annotation on the parameter is much more clear which parameter it applies to, avoids the need for duplicating the parameter name, and is much easier to line up when the number of parameters annotated grows